blueocean-i18n/src/main/java/io/jenkins/blueocean/i18n/BlueI18n.java

```diff
@@ -0,0 +1,394 @@
+ * THE SOFTWARE.
+ */
+package io.jenkins.blueocean.i18n;
+
+import hudson.Extension;
+import hudson.PluginWrapper;
+import hudson.model.RootAction;
+import hudson.util.HttpResponses;
+import jenkins.model.Jenkins;
+import net.sf.json.JSONObject;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+import org.kohsuke.stapler.HttpResponse;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+
+import javax.annotation.CheckForNull;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.MissingResourceException;
+import java.util.ResourceBundle;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Pattern;
+
+/**
+ * Internationalization REST (ish) API for Blue Ocean.
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+@Extension
+@Restricted(NoExternalUse.class)
+public class BlueI18n implements RootAction {
+
+    private static final Logger LOGGER = Logger.getLogger(BlueI18n.class.getName());
+
+    /**
+     * Failed lookup cache entry.
+     */
+    private static final JSONObject BUNDLE_404 = new JSONObject();
+
+    /**
+     * Bundle cache.
+     */
+    private Map<BundleParams, BundleCacheEntry> bundleCache = new ConcurrentHashMap<>();
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getIconFileName() {
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getDisplayName() {
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getUrlName() {
+        return "blueocean-i18n";
+    }
+
+    /**
+     * Get a localised resource bundle.
+     * <p>
+     * URL: {@code blueocean-i18n/$PLUGIN_NAME/$PLUGIN_VERSION/$BUNDLE_NAME/$LOCALE} (where {@code $LOCALE} is optional).
+     *
+     * @param request The request.
+     * @return The JSON response.
+     */
+    public HttpResponse doDynamic(StaplerRequest request) {
+        BundleParams bundleParams = getBundleParameters(request.getRestOfPath());
+
+        if (bundleParams == null) {
+            return HttpResponses.errorJSON("All mandatory bundle identification parameters not specified: '$PLUGIN_NAME/$PLUGIN_VERSION/$BUNDLE_NAME' (and optional $LOCALE).");
+        }
+
+        try {
+            Locale locale = bundleParams.getLocale();
+
+            if (locale == null) {
+                locale = request.getLocale();
+            }
+
+            BundleCacheEntry bundleCacheEntry = bundleCache.get(bundleParams);
+
+            JSONObject bundle;
+            if (bundleCacheEntry == null) {
```

Comment:
bundleCache is ConcurrentHashMap so thread safety is taken care. However if you use guava cache, it might have better caching behavior/functionality. Not a big deal though.